src/cljs/harja/ui/napit.cljs

```diff
@@ -73,21 +91,16 @@
              :class    (if (or @kysely-kaynnissa? (:disabled asetukset))
                          (str luokka " disabled")
                          luokka)
-             :on-click #(do
-                          (.preventDefault %)
-                          (.stopPropagation %)
-                          (reset! kysely-kaynnissa? true)
-                          (reset! nayta-virheviesti? false)
-                          (go (let [tulos (<! (kysely))]
-                                (reset! kysely-kaynnissa? false)
-                                (when kun-valmis (kun-valmis tulos))
-                                (if (not (k/virhe? tulos))
-                                  (when kun-onnistuu
-                                    (kun-onnistuu tulos))
-                                  (do
-                                    (log "VIRHE PALVELINKUTSUSSA!" (pr-str tulos))
-                                    (reset! nayta-virheviesti? true)
-                                    (when kun-virhe (kun-virhe tulos)))))))
+             :on-click #(suorita-kysely %)
+             :on-key-down #(do
+                             (when (dom/enter-nappain? %)
+                               (suorita-kysely %)
+                               (when siirra-fokus
+                                 (r/after-render (fn [] (some-> js/document (.getElementById (str siirra-fokus)) .focus)))))
+                             (when (or (and (dom/tab-nappain-ilman-shiftia? %) esta-tab-eteenpain?)
+                                     (and (dom/tab+shift-nappaimet? %) esta-tab-taaksepain?))
+                               (.preventDefault %)
+                               (.stopPropagation %)))
```

Comment:
Tässä setissä hieman koodin kanssa epäloogisuutta.

```clojure
(when
   (dom/enter-nappain? %)
   (suorita-kysely %)
   (when siirra-fokus 
      ....) ; <- nestattu when
```

Minä luulen että tämän koko himmelin voi tehdä condition lauseella. Object oriented kielistä varmasti tuttu switch? Tämä on sama homma.
Jos haluat että condition jatkuu (eli suoritetaan kaikki conditiot), voi käyttää threading macroa ->
Elikkäs, näin:

```clojure
:on-key-down #(cond->
	  ;; Enter näppäintä painettiin
	  (dom/enter-nappain? %)
	  (suorita-kysely %)
	  ;; Fokus
	  siirra-fokus
	  (r/after-render (fn []
				(some-> js/document (.getElementById (str siirra-fokus)) .focus)))
	  ;;  Tabitus
	  (or
		(and (dom/tab-nappain-ilman-shiftia? %) esta-tab-eteenpain?)
		(and (dom/tab+shift-nappaimet? %) esta-tab-taaksepain?))
	  (do
		(.preventDefault %)
		(.stopPropagation %)))
```

Jos haluat, että tuo koodin suoritus loppuu ensimmäiseen true statementtiin, joka olisi tässä keississä (dom/enter-nappain? %) ..
Voit ottaa threading macron pois. HOX en ole testannut yllä olevaa koodia, pitäisi varmistaa että tuo toimii oikein

Reply:
Näissä olisi mielestäni tärkeää dokumentoida palvelinkutsu-nappi ja napin kuvaukseen miten uusia arvoja käytetään käytännössä jotta tulisi käytettyä myös jatkossa.